### DISCUSSION TITLE
Publicodes : équivalence module CJS / ESM

### DISCUSSION THREAD

**Discussion by laem** [publicodes/publicodes#70] [Q&A]
Je remarque que les exports CJS/ESM sont bien différents.

https://github.com/betagouv/mon-entreprise/blob/master/publicodes/core/source/index.ts

https://github.com/betagouv/mon-entreprise/blob/master/publicodes/core/esm/index.js

En particulier, il me faut serializeUnit, qui est dans CJS mais pas dans ESM.

J'ai donc pour l'instant fait deux imports, l'un `import` et l'autre `require` exceptionnellement. Je crois que c'est le passage à Webpack 5 betagouv/mon-entreprise#1491 qui a forcé l'utilisation de la version ESM.

@johangirod pourquoi avoir deux périmètres différents pour ces exports ? C'est documenté quelque part ?

**Comment by johangirod** [comment]
Houlà, non, bien au contraire, il faudrait qu'ils soient exactement les mêmes. J'imagine que je n'ai pas tout ajouté à l'ESM dans l'idée de restreindre l'API officielle : par exemple, il n'y aucun besoin d'exporter `capitalize0`. En revanche pour `serializeUnit`, c'est effectivement un bon point.

Il faudrait cependant trouver un moyen esm/index.js à partir de source/index.js... Ou alors encore mieux, utiliser ESM nativement de partout et compiler après coup en CJS au besoin. Mais j'avais un peu tenté et je m'étais mangé pas mal de difficultés.

En attendant, l'idée est d'aboutir à une API assez stable et restreinte pour la v1, la question des unités est pile poil dedans.

Sur la question "pourquoi deux exports ?" voir l'article suivant :
https://redfin.engineering/node-modules-at-war-why-commonjs-and-es-modules-cant-get-along-9617135eeca1

En attendant, tu peux ajouter l'export serializeUnit dans esm si tu en as besoin

**Comment by laem** [comment]
Merci pour les détails. Le double import/require ne me gène pas trop pour l'instant :)
